// @sismo-core/crypto/lib/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ethers = require('ethers');
var circomlibjs = require('circomlibjs');

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

var babyJub;
function buildBabyjub() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!babyJub) return [3 /*break*/, 2];
                    return [4 /*yield*/, circomlibjs.buildBabyjub()];
                case 1:
                    babyJub = _a.sent();
                    _a.label = 2;
                case 2: return [2 /*return*/, babyJub];
            }
        });
    });
}

var eddsa;
function buildEddsa() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!eddsa) return [3 /*break*/, 2];
                    return [4 /*yield*/, circomlibjs.buildEddsa()];
                case 1:
                    eddsa = _a.sent();
                    _a.label = 2;
                case 2: return [2 /*return*/, eddsa];
            }
        });
    });
}

var EddsaAccount = /** @class */ (function () {
    function EddsaAccount(pubKey, privKey, eddsa, babyjub) {
        this.pubKey = pubKey;
        this.privKey = privKey;
        this.eddsa = eddsa;
        this.babyJub = babyjub;
    }
    EddsaAccount.generateFromSeed = function (seed) {
        return __awaiter(this, void 0, void 0, function () {
            var eddsa, babyjub, privKey, pubKey;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, buildEddsa()];
                    case 1:
                        eddsa = _a.sent();
                        return [4 /*yield*/, buildBabyjub()];
                    case 2:
                        babyjub = _a.sent();
                        privKey = Buffer.from(ethers.BigNumber.from(seed).toHexString().slice(2), "hex");
                        pubKey = eddsa.prv2pub(privKey);
                        return [2 /*return*/, new EddsaAccount(pubKey, privKey, eddsa, babyjub)];
                }
            });
        });
    };
    EddsaAccount.prototype.sign = function (msg) {
        var msgF = this.babyJub.F.e(ethers.BigNumber.from(msg));
        var signature = this.eddsa.signPoseidon(this.privKey, msgF);
        return [
            ethers.BigNumber.from(this.babyJub.F.toObject(signature.R8[0])),
            ethers.BigNumber.from(this.babyJub.F.toObject(signature.R8[1])),
            ethers.BigNumber.from(signature.S)
        ];
    };
    EddsaAccount.verify = function (msg, signature, pubKey) {
        return __awaiter(this, void 0, void 0, function () {
            var babyJub, eddsa, msgF, sigPoseidon, pubKeyPoseidon;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, buildBabyjub()];
                    case 1:
                        babyJub = _a.sent();
                        return [4 /*yield*/, buildEddsa()];
                    case 2:
                        eddsa = _a.sent();
                        msgF = babyJub.F.e(ethers.BigNumber.from(msg));
                        sigPoseidon = {
                            R8: [
                                babyJub.F.e(ethers.BigNumber.from(signature[0])),
                                babyJub.F.e(ethers.BigNumber.from(signature[1]))
                            ],
                            S: ethers.BigNumber.from(signature[2]).toBigInt()
                        };
                        pubKeyPoseidon = [
                            babyJub.F.e(ethers.BigNumber.from(pubKey[0])),
                            babyJub.F.e(ethers.BigNumber.from(pubKey[1])),
                        ];
                        return [2 /*return*/, eddsa.verifyPoseidon(msgF, sigPoseidon, pubKeyPoseidon)];
                }
            });
        });
    };
    EddsaAccount.prototype.getPubKey = function () {
        return [
            ethers.BigNumber.from(this.babyJub.F.toObject(this.pubKey[0])),
            ethers.BigNumber.from(this.babyJub.F.toObject(this.pubKey[1]))
        ];
    };
    return EddsaAccount;
}());

var poseidon;
function buildPoseidon() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!poseidon) return [3 /*break*/, 2];
                    return [4 /*yield*/, circomlibjs.buildPoseidon()];
                case 1:
                    poseidon = _a.sent();
                    _a.label = 2;
                case 2: return [2 /*return*/, function (inputs) {
                        var hash = poseidon(inputs.map(function (x) { return ethers.BigNumber.from(x); }));
                        var hashStr = poseidon.F.toString(hash);
                        return ethers.BigNumber.from(hashStr);
                    }];
            }
        });
    });
}

var SNARK_FIELD = ethers.BigNumber.from("21888242871839275222246405745257275088548364400416034343698204186575808495617");

function keccak256(inputs) {
    return ethers.BigNumber.from(ethers.ethers.utils.keccak256(ethers.ethers.utils.solidityPack(["uint256", "uint256"], [inputs[0], inputs[1]])));
}

exports.EddsaAccount = EddsaAccount;
exports.SNARK_FIELD = SNARK_FIELD;
exports.buildPoseidon = buildPoseidon;
exports.keccak256 = keccak256;
